RESOURCE_ACCESS_CONTROL_FOR_PLUGINS.md

```diff
@@ -0,0 +1,691 @@
+
+
+### **Example: Private Resource**
+
+To keep a resource fully private:
+
+```json
+{
+  "share_with": {}
+}
+```
+
+Since no entities are listed, the resource is accessible **only by its creator and super-admins**.
+
+**This is the default state for all new resources.**
+
+---
+
+## **7. User Setup**
+
+To enable users to interact with the **Resource Sharing and Access Control** feature, they must be assigned the appropriate cluster permissions along with resource-specific access.
+
+### **Required Cluster Permissions**
+Users must be assigned the following **cluster permissions** in `roles.yml`:
+
+- **Plugin-specific cluster permissions** → Required to interact with the plugin’s APIs.
+
+#### **Example Role Configurations**
+```yaml
+sample_full_access:
+  cluster_permissions:
+    - 'cluster:admin/sample-resource-plugin/*'
+
+sample_read_access:
+  cluster_permissions:
+    - 'cluster:admin/sample-resource-plugin/get'
+```
+
+### **User Access Rules**
+1. **Users must have the required cluster permissions**
+    - Even if a resource is shared with a user, they **cannot access it** unless they have the **plugin’s cluster permissions**.
+
+2. **Granting plugin API permissions does not automatically grant resource access**
+    - A resource must be **explicitly shared** with the user.
+    - **Or, the user must be the resource owner.**
+
+3. **No index permissions are required**
+    - Access control is **handled at the cluster level**.
+    - The `.opensearch_resource_sharing` index and the resource indices are protected under system index security.
+
+
+### **Summary**
+| **Requirement** | **Description**                                                                       |
+|---------------|---------------------------------------------------------------------------------------|
+| **Plugin API Permissions** | Users must also have relevant plugin API cluster permissions.                         |
+| **Resource Sharing** | Access is granted only if the resource is shared with the user or they are the owner. |
+| **No Index Permissions Needed** | The `.opensearch_resource_sharing` index and resource indices are system-protected.   |
+
+
+---
+
+## **8. Restrictions**
+1. At present, **only resource owners can share/revoke access** to their own resources.
+    - **Super admins** can manage access for any resource.
+2. **Resources must be stored in a system index**, and system index protection **must be enabled**.
+    - **Disabling system index protection** allows users to access resources **directly** if they have relevant index permissions.
+
+---
+
+## **9. REST APIs Introduced by the Security Plugin**
+
+In addition to client methods, the **Security Plugin** introduces new **REST APIs** for managing resource access when the feature is enabled. These APIs allow users to **verify, grant, revoke, and list access** to resources.
+
+### **1. Verify Access**
+- **Endpoint:**
+  ```
+  POST /_plugins/_security/resources/verify_access
```

Comment:
In what instances would plugins call this REST API vs using the service class that has a similar API?

One API I could see useful for powering frontend UX for sharing would be an API to get the sharing info of a current resource.

i.e. in Quip if I own a document, I can click on the share button and it will show me who the document has been shared with an at what access levels.

Reply:
okay, I will remove the code.